lib/srv/discovery/discovery.go

```diff
@@ -1327,8 +1323,90 @@ func (s *Server) getAllGCPServerFetchers() []server.Fetcher {
 	return allFetchers
 }
 
+// acquireDiscoveryGroup tries to acquire a lock if the Service has a DiscoveryGroup
+// It will retry using an exponential backoff algorithm.
+func (s *Server) acquireDiscoveryGroup() error {
+	if s.DiscoveryGroup == "" {
+		return nil
+	}
+
+	retry, err := retryutils.NewRetryV2(retryutils.RetryV2Config{
+		First:  0,
+		Driver: retryutils.NewExponentialDriver(defaults.HighResPollingPeriod),
+		Max:    defaults.LowResPollingPeriod,
+		Jitter: retryutils.NewHalfJitter(),
+		Clock:  s.clock,
+	})
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	var lease *services.SemaphoreLock
+	for range retry.After() {
+		retry.Inc()
+		s.Log.Debugf("Discovery service is trying to acquire lock for DiscoveryGroup %q", s.DiscoveryGroup)
+		lease, err = s.tryAcquireDiscoveryGroupLease()
+		if err != nil {
+			if !strings.Contains(err.Error(), teleport.MaxLeases) {
+				return trace.Wrap(err)
+			}
+			s.Log.Debugf("Discovery service is waiting on DiscoveryGroup %q lock: %v", s.DiscoveryGroup, err)
+			continue
+		}
+
+		break
+	}
+
+	go func() {
+		for {
+			select {
+			case <-lease.Renewed():
+				continue
+			case <-lease.Done():
+				s.Log.WithError(lease.Wait()).Warnf("DiscoveryGroup %q lock was lost, stopping discovery service", s.DiscoveryGroup)
+				s.Stop()
+				return
+			}
+		}
+	}()
+
+	return nil
+}
+
+func (s *Server) tryAcquireDiscoveryGroupLease() (*services.SemaphoreLock, error) {
+	lease, err := services.AcquireSemaphoreLock(s.ctx, services.SemaphoreLockConfig{
+		Service: s.Config.AccessPoint,
+		Expiry:  time.Minute,
+		Params: types.AcquireSemaphoreRequest{
+			SemaphoreKind: types.SemaphoreKindDiscoveryServiceGroup,
+			SemaphoreName: s.DiscoveryGroup,
+			MaxLeases:     1,
+			Holder:        s.ServerID,
+		},
+		Clock: s.clock,
+	})
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	s.relaseGroupMu.Lock()
+	defer s.relaseGroupMu.Unlock()
+	s.releaseGroupLockFn = func() {
+		lease.Stop()
+		if err := lease.Wait(); err != nil {
+			s.Log.WithError(err).Warn("DiscoveryGroup semaphore cleanup")
+		}
+	}
+
+	return lease, nil
+}
+
 // Start starts the discovery service.
 func (s *Server) Start() error {
+	if err := s.acquireDiscoveryGroup(); err != nil {
+		return trace.Wrap(err)
+	}
+
 	if s.ec2Watcher != nil {
 		go s.handleEC2Discovery()
 		go s.reconciler.run(s.ctx)
```

Comment:
```suggestion
func (s *Server) Start() error {
       for {
	err := s.runWhileAcquiringDiscoveryGroup(s.start)
	switch{
	case error.Is(...) /* lease was lost*/:
	     continue
	default:
	     return trace.Wrap(err) 
	}
	
      }
}


func (s *Server) start(ctx context.Context) error {
	if s.ec2Watcher != nil {
		go s.handleEC2Discovery()
		go s.reconciler.run(ctx)
		....
```

Reply:
I'm not sure the service is ready to be started multiple times.
The `runWhileAcquiringDiscoveryGroup` would also need to wait for the service to clean up. Eg watchers, fetchers, installers and reconcilers would need to stop before we can proceed.